lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -252,6 +270,32 @@ impl Client {
                 debug!("sending device IO response");
                 Ok(resp)
             }
+            MajorFunction::IRP_MJ_CREATE => {
+                let rdp_req = ServerCreateDriveRequest::decode(device_io_request, payload)?;
+                debug!("got: {:?}", rdp_req);
+
+                // Send a TDP Shared Directory Info Request
+                (self.tdp_sd_info_request)(SharedDirectoryInfoRequest::from(rdp_req.clone()))?;
+
+                // Add a TDP Shared Directory Info Response handler to the handler cache.
+                // When we receive a TDP Shared Directory Info Response with this completion_id,
+                // this handler will be called.
+                self.pending_sd_info_resp_handlers.insert(
```

Comment:
I'm curious to see what a more experienced Rustacean thinks about this approach with caching dynamic callbacks.

There might be a more "Rusty" way to establish a state machine based on an enum.


cc @xacrimon @espadolini @timothyb89

Reply:
Storing boxed callbacks here does feel unidiomatic. Boxed callbacks are typically the right solution when you have functions with different logic going on that call need go be executed from the same place.

Here we have homogeneous logic, we simply use the boxed closure abstraction to capture some state variables that we consume in the callback. I'd rather prefer if we moved the logic in the boxed callback out in a function that takes all values as parameters instead of capturing them in a closure. Then we can replace this callback map with `HashMap<u32, State>` and simply pass the state to the function instead, the state should contain all of the variables that we currently capture in the closure.

This doesn't seem to warrant a full state machine, we just need to store some parameters for the callback we run when we receive the response.

Reply:
If we need to keep track of state between more than a simple request/response pair, we would need proper state machines which are probably best implemented as enums with one variant per possible state and let these state enums be modified by callbacks method on client that are generic across connections.

Reply:
In this specific case, it seems like it would mostly involve storing the rdp_req variable and a connection ID inside of a state struct we store here that is then used when we receive a TDP Shared Directory Info Response message.

Reply:
I thought about `HashMap<u32, State>`, one reason I went with this approach instead is that it reads in the same way you think about the algorithm (as I try to relay in the PR description). If I were to go with state, it would read like

```rust
MajorFunction::IRP_MJ_CREATE => {
  let rdp_req = ServerCreateDriveRequest::decode(device_io_request, payload)?;
  debug!("got: {:?}", rdp_req);

  (self.tdp_sd_info_request)(SharedDirectoryInfoRequest::from(rdp_req.clone()))?;


  self.pending_sd_info_resp_handlers.insert(
      rdp_req.device_io_request.completion_id,
      State::new(rdp_req)
  );
  
  Ok(vec![])
}
```

It ends abruptly, it's not clear where the logic that actually deals with the bulk of the `IRP_MJ_CREATE` logic is coming from. I could of course add a comment telling the reader where to go, but that's less elegant than my current solution where they essentially just continue to follow the code top to bottom. (Keep in mind that the `// TODO(isaiah): see https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L207` is doing a lot of work here. This will wind up being a quite complex bit of code with some code paths requiring additional TDP Request/Responses. I think it would end up being very difficult to follow/maintain if done with comments).

Additionally, `Shared Directory Info Response`s aren't always handled in the same manner. Here we will handle them one way, but we will also need a handler for other contexts (e.g. dealing with `IRP_MJ_QUERY_INFORMATION`). This means that we will need to add an `enum` to the `State` that determines which handler will be called. Which works fine, but is more steps and if statements than just grabbing the handler from the `HashMap` and calling it.

I am of course open to being proven wrong about my desire for these properties, or of being shown a more idiomatic way to get the same properties, but currently still feel strongly that what I have is superior to `HashMap<u32, State>`.

Reply:
If so, I'd be happy to keep this, at least until the needs for this logic change.